File: 2021/day_1/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func readFile(fileName string) ([]string, error) {
	data, err := os.ReadFile(fileName)
	if err != nil {
		return nil, err
	}
	parsed := string(data)
	split := strings.Split(parsed, "\n")
	return split, nil
}

func parseToInt(s []string) []int {
	numbered_lines := []int{}
	for _, line := range s {
		if line == "" {
			continue
		}
		parsed, err := strconv.Atoi(line)
		if err != nil {
			panic(err)
		}
		numbered_lines = append(numbered_lines, parsed)
	}
	return numbered_lines

}

func checkIncreasing(l []int) []bool {
	previous := 10000000
	result := []bool{}
	for _, current := range l {
		result = append(result, previous < current)
		previous = current
	}
	return result
}

func checkIncreasingSliding(l []int) []bool {
	window := [3]int{0, 0, 0}
	result := []bool{}
	for i, v := range l {
		if i >= 3 {
			result = append(result, v > window[0])
		}
		window = [3]int{window[1], window[2], v}
	}
	return result
}

func countTrue(l []bool) int {
	result := 0
	for _, value := range l {
		if value == true {
			result += 1
		}
	}
	return result
}

func main() {
	filename := "input.txt"
	lines, err := readFile(filename)
	if err != nil {
		lines, err = readFile(fmt.Sprintf("2021/day_1/%s", filename))
		if err != nil {
			panic(err)
		}
	}
	numbers := parseToInt(lines)
	// increasing := checkIncreasing(numbers) // Part 1
	increasing := checkIncreasingSliding(numbers) // Part 2
	countIncreasing := countTrue(increasing)
	fmt.Printf("There are %d increasing values in the list", countIncreasing)
}

File: 2021/day_1/main.rs
use std::fs;

fn read_file(path: &str) -> Vec<String> {
    let parsed = fs::read_to_string(path).expect("Could not read file");
    let split_lines = parsed.lines();
    split_lines.map(|s| s.to_string()).collect()
}

fn parse_to_int(lines: &Vec<String>) -> Vec<i32> {
    let mut result = Vec::<i32>::new();
    for value in lines {
        result.push(value.parse().unwrap());
    }
    result
}

fn check_increasing(values: &Vec<i32>) -> Vec<bool> {
    let mut previous: i32 = 100000000;
    let mut result = Vec::<bool>::new();
    for value in values {
        result.push(*value > previous);
        previous = *value;
    }
    result
}

fn check_increasing_sliding(values: &Vec<i32>) -> Vec<bool> {
    let mut window = (0, 0, 0);
    let mut result = Vec::<bool>::new();
    for (i, value) in values.iter().enumerate() {
        if i >= 3 {
            result.push(*value > window.0);
        }
        window = (window.1, window.2, *value);
    }
    result
}

fn count_true(values: &Vec<bool>) -> i32 {
    values.iter().fold(0, |c, item| c + (*item as i32))
}

fn main() {
    let filename = "2021/day_1/input.txt";
    let lines = read_file(filename);
    let numbers = parse_to_int(&lines);
    // let increasing = check_increasing(&numbers); // Part 1
    let increasing = check_increasing_sliding(&numbers); // Part 2
    let count = count_true(&increasing);
    println!("There are {:?} increasing steps", count);
}
